### ISSUE TITLE
pymbar 4.0 and API changes?

### ISSUE THREAD

**Issue by mrshirts** [alchemistry/alchemlyb#204]
We are moving towards a pymbar 4.0 release in the next while (exact timing to be determined, could be on the order of weeks).   The current state of the 4.0 code is at https://github.com/choderalab/pymbar/tree/pymbar4/

The main points are:
-  pymbar returns will be dictionary based, rather than tuple bases, making it easier to access
- There will be more control over adaptivity of results and workflow.
- bootstrapped errors (in addition to analytical errors)
- functions use snake_case naming. 
- output uses the logging module.
- 
The API is not yet fully documented, but you can see the calls in https://github.com/choderalab/pymbar/tree/pymbar4/examples, for example.  We will try try to get something more concrete soon.

We are also looking at acceleration with jax.  Are there advantages to making this optional, rather than required, to fit into alchemlyb workflows better?  Are there issues with adding jax dependencies?

How can we help to make incorporation easier?

Now is also a good time for any other suggestions.

**Comment by orbeckst** [comment]
Not having jax a required dependency would be good.

**Comment by mrshirts** [comment]
Is there a way to make an optional dependency?  We could release 2 packages, I'm not sure how complicated that would be.  Fortunately the amount of different code is low.

**Comment by xiki-tempula** [comment]
@mrshirts Many thanks for the work. I wonder the pymbar 4.0 is API stable now? So I could have a test to see what are the differences that needs to be made in alchemlyb.

**Comment by mrshirts** [comment]
Essentially, yes.  The key changes are returns as dictionaries, and
standardized snake_case command names. I might add additional return values
for bootstrapping, but that would not break things (and is one of the
reasons to switch to a dictionary).  The common commands should not change
either.

Apologies that the documentation for the API is not yet done, that's one of
the tasks for the 4.0 release!

**Comment by xiki-tempula** [comment]
@mrshirts Thank you for answering. I wonder what is the plan for the current pymbar 3? Would it be like py2/py3 where pymbar 3 will stay one for some time or everything will go to pymbar 4?

**Comment by mrshirts** [comment]
There is not yet consensus on this.  I need feedback from a few more people.  If you have a suggestion, let me know.  Given the difficulty of maintaining two branches given personnel (mostly just me), probably not for very long.

**Comment by xiki-tempula** [comment]
@mrshirts Thanks. I was looking at the code and thinking of it. If the result is returned as dictionary, can we assume that the dictionary will **definitely** have certain keywords with defined output or it is a more fluid situation?

**Comment by mrshirts** [comment]
Yes, we are not planning on the keywords changing.  Possibly some could be added, but the plan is for them not to change anymore (on the forseeable timeline).

**Comment by orbeckst** [comment]
For alchemlyb we will not be able to support pymbar 3 and 4 in parallel — that's just too complicated. 

I propose that we are aiming for an **alchemlyb 1.0 release that supports pymbar 3.x.** (We should to do a 1.0 anyway...)

We can then start work on **alchemlyb 2.0 with pymbar 4.x.** In this way it is very clear what is supported and we have at least the option to backport 2.x fixes to 1.x if really necessary (although that will mostly depend on someone offering to do the work...).
